File: web/preview/diagnostics.js
function lineLabel(item, sourceLineOffset) {
  const line = item.span?.start_line ?? item.span?.line;
  if (line) return `Line ${Math.max(1, line - sourceLineOffset)}`;
  if (item.span?.first) return `Byte ${item.span.first}`;
  return "Document";
}

export function createDiagnosticsView(root) {
  const toggle = root.querySelector("#diagnostics-toggle");
  const list = root.querySelector("#diagnostics-list");
  const count = root.querySelector("#diagnostics-count");
  let expanded = false;

  function setExpanded(next) {
    expanded = next;
    toggle.setAttribute("aria-expanded", String(next));
    list.hidden = !next;
  }

  toggle.addEventListener("click", () => setExpanded(!expanded));

  return function renderDiagnostics(items = [], sourceLineOffset = 0) {
    list.replaceChildren();
    const errors = items.filter((item) => !item.severity || item.severity === "error");
    const warnings = items.filter((item) => item.severity === "warning");
    root.dataset.tone = errors.length ? "error" : warnings.length ? "warning" : "ready";
    count.textContent = String(items.length);

    items.forEach((item) => {
      const row = document.createElement("li");
      row.className = "diagnostic-item";
      row.dataset.severity = item.severity || "error";
      const location = document.createElement("span");
      location.className = "diagnostic-location";
      location.textContent = lineLabel(item, sourceLineOffset);
      const message = document.createElement("span");
      message.className = "diagnostic-message";
      const code = document.createElement("b");
      code.textContent = item.code || "DIAGNOSTIC";
      const detail = document.createElement("span");
      detail.textContent = item.message || "Unknown formatter issue";
      message.append(code, detail);
      row.append(location, message);
      list.append(row);
    });

    if (items.length && !expanded) setExpanded(true);
    if (!items.length && expanded) setExpanded(false);
    return { errors: errors.length, warnings: warnings.length };
  };
}

File: src/tm_u220/core/diagnostics.lua
local M = {}

function M.new(code, message, span, severity)
    return {
        code = assert(code),
        message = assert(message),
        severity = severity or "error",
        span = span or {},
    }
end

function M.has_errors(items)
    for _, item in ipairs(items or {}) do
        if item.severity == "error" then
            return true
        end
    end
    return false
end

function M.format(item)
    local where = ""
    item.span = item.span or {}
    if item.span.start_line then
        where = "line " .. item.span.start_line .. ": "
    elseif item.span.line then
        where = "line " .. item.span.line .. ": "
    elseif item.span.first then
        where = "byte " .. item.span.first .. ": "
    end
    return string.format("%s[%s] %s%s", item.severity, item.code, where, item.message)
end

return M
